src/sentry/integrations/gitlab/client.py

```diff
@@ -126,11 +119,28 @@ def refresh_auth(self):
         return self.identity.get_provider().refresh_identity(
             self.identity,
             refresh_token_url="{}{}".format(
-                self.metadata["base_url"], GitLabApiClientPath.oauth_token
+                self.base_url.rstrip("/"), GitLabApiClientPath.oauth_token
             ),
             verify_ssl=self.metadata["verify_ssl"],
         )
 
+    def request(self, *args: Any, **kwargs: Any):
+        try:
+            return super().request(*args, **kwargs)
+        except ApiUnauthorized as e:
+            if self.is_refreshing_token:
+                raise e
+
+            self.is_refreshing_token = True
+            self.new_identity = self._refresh_auth()
+
+            response = super().request(*args, **kwargs)
+
+            self.is_refreshing_token = False
+            self.new_identity = None
+
+            return response
```

Comment:
`def request()` isn't silo specific, but we don't want any of the refresh code being executed on region silos. I'm not sure if there's a more consistent way to refresh (rather than on 403s) but you'll need to contain that to the control silo exclusively.

A good way to do that is if we store the expiration of tokens, using that rather than a 403 as signal to refresh (that is what was done for GitHub)